File: Python/Combination Sum.py
class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum(self, candidates, target):
        ret=[]
        candidates.sort() 
        for i,item in enumerate(candidates):
            if target>item:
                subSet=self.combinationSum(candidates[i:],target-item) 
                if subSet:
                    ret+=[[item]+lis for lis in subSet]
            elif target==item:
                ret+=[[item]]
            else:
                break
        return ret
s = Solution()
print s.combinationSum([2,3,6,7], 7)

File: Java/src/Combination Sum.java
public class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ret = new ArrayList<>();
        Arrays.sort(candidates);
        for(int i = 0; i<candidates.length; i++){
            if (candidates[i] < target){
                 List<List<Integer>> subSet = combinationSum(Arrays.copyOfRange(candidates,i,candidates.length), target- candidates[i]);
                 
                 for(List<Integer> list : subSet){
                    list.add(0,candidates[i]);
                    ret.add(list);
                 }
            }
            else if (candidates[i] == target){
                List<Integer> newList = new ArrayList<Integer>();
                newList.add(candidates[i]);
                ret.add(newList);
            }
            else{
                break;
            }
        }
        return ret;
    }
}